asmcomp/cmmgen.ml

```diff
@@ -2489,9 +2500,9 @@ let emit_constant_closure symb fundecls cont =
             emit_others (pos + 4) rem in
       Cint(black_closure_header (fundecls_size fundecls)) ::
       Cdefine_symbol symb ::
-      if f1.arity = 1 then
+      if f1.arity = 1 || f1.arity = 0 then
         Csymbol_address f1.label ::
-        Cint 3n ::
+        Cint(Nativeint.of_int (f1.arity lsl 1 + 1)) ::
```

Comment:
Actually, I would argue the existing code is more in line with the typical code used to tag an integer.